ibis/expr/types/relations.py

```diff
@@ -659,12 +659,21 @@ def bind(self, *args: Any, **kwargs: Any) -> tuple[Value, ...]:
             A tuple of bound values
         """
         values = self._fast_bind(*args, **kwargs)
-        # dereference the values to `self`
-        dm = DerefMap.from_targets(self.op())
+        dm = None  # delay creating a dereference map until one is definitely needed
+        this_table_relations_set = {self.op()}
         result = []
         for original in values:
-            value = dm.dereference(original.op()).to_expr()
-            value = value.name(original.get_name())
+            rels = original.op().relations
+            if len(rels) and rels != this_table_relations_set:
+                # the expression needs dereferencing as it has relations (i.e. is not a literal) and is not bound to self
+                if dm is None:
+                    # create a dereference map as one has not been created yet
+                    dm = DerefMap.from_targets(self.op())
+                # dereference the values to `self`
+                value = dm.dereference(original.op()).to_expr()
```

Comment:
It might be nicer to construct the derefmap's internal state lazily. What I'm thinking of:
1. `DerefMap.from_targets` creates a `DerefMap` instance with the targets stored, but without actually traversing the values backwards, hence it would be an empty dictionary.
2. `dm.dereference()` would check `rels != this_table_relations_set` first, and only fill the dereference mapping when needed.

Basically the optimization could be pushed down entirely to the dereference mapping.

Reply:
This work is a little bit more involved so we can defer it to follow-up PRs.

Reply:
Let's do the refactor in a follow-up, I can take that one!